src/hotspot/cpu/ppc/gc/shenandoah/shenandoahBarrierSetAssembler_ppc.cpp

```diff
@@ -0,0 +1,1017 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "gc/shared/gcArguments.hpp"
+#include "gc/shared/gc_globals.hpp"
+#include "macroAssembler_ppc.hpp"
+#include "precompiled.hpp"
+#include "asm/macroAssembler.inline.hpp"
+#include "gc/shenandoah/shenandoahBarrierSet.hpp"
+#include "gc/shenandoah/shenandoahBarrierSetAssembler.hpp"
+#include "gc/shenandoah/shenandoahForwarding.hpp"
+#include "gc/shenandoah/shenandoahHeap.hpp"
+#include "gc/shenandoah/shenandoahHeap.inline.hpp"
+#include "gc/shenandoah/shenandoahHeapRegion.hpp"
+#include "gc/shenandoah/shenandoahRuntime.hpp"
+#include "gc/shenandoah/shenandoahThreadLocalData.hpp"
+#include "gc/shenandoah/heuristics/shenandoahHeuristics.hpp"
+#include "interpreter/interpreter.hpp"
+#include "runtime/sharedRuntime.hpp"
+#include "runtime/thread.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "vm_version_ppc.hpp"
+
+#ifdef COMPILER1
+
+#include "c1/c1_LIRAssembler.hpp"
+#include "c1/c1_MacroAssembler.hpp"
+#include "gc/shenandoah/c1/shenandoahBarrierSetC1.hpp"
+
+#endif
+
+#define __ masm->
+
+void ShenandoahBarrierSetAssembler::satb_write_barrier(MacroAssembler *masm,
+                                                       Register base, RegisterOrConstant ind_or_offs,
+                                                       Register tmp1, Register tmp2, Register tmp3,
+                                                       MacroAssembler::PreservationLevel preservation_level) {
+  if (ShenandoahSATBBarrier) {
+    __ block_comment("satb_write_barrier (shenandoahgc) {");
+    satb_write_barrier_impl(masm, 0, base, ind_or_offs, tmp1, tmp2, tmp3, preservation_level);
+    __ block_comment("} satb_write_barrier (shenandoahgc)");
+  }
+}
+
+void ShenandoahBarrierSetAssembler::iu_barrier(MacroAssembler *masm,
+                                               Register val,
+                                               Register tmp1, Register tmp2,
+                                               MacroAssembler::PreservationLevel preservation_level,
+                                               DecoratorSet decorators) {
+  // IU barriers are also employed to avoid resurrection of weak references,
+  // even if Shenandoah does not operate in incremental update mode.
+  if (ShenandoahIUBarrier || ShenandoahSATBBarrier) {
```

Comment:
Sharing the code for IU and SATB sounds like a good idea, but one needs to be careful. `ShenandoahBarrierSetC1::iu_barrier` only works with `ShenandoahIUBarrier`, so this trick can't be used in C1.
It's a bit confusing, but I'm ok with this version. At least, I don't have any better suggestion at the moment.